packages/sources/blocksize-capital-test/src/endpoint/price.ts

```diff
@@ -0,0 +1,145 @@
+import {
+  WebsocketReverseMappingTransport,
+  WebsocketTransportGenerics,
+  WebSocketRawData,
+} from '@chainlink/external-adapter-framework/transports/websocket'
+import {
+  PriceEndpoint,
+  PriceEndpointParams,
+  PriceEndpointInputParameters,
+} from '@chainlink/external-adapter-framework/adapter'
+import { customSettings } from '../config'
+
+const logger = makeLogger('BlocksizeCapitalWebsocketEndpoint')
+interface BaseMessage {
+  jsonrpc: string
+  id?: string | number | null
+}
+
+export interface Message extends BaseMessage {
+  method: 'vwap'
+  params: {
+    updates: {
+      ticker: string
+      price?: number
+      size?: number
+      volume?: number
+      ts: number
+    }[]
+  }
+}
+
+const inputParameters: PriceEndpointInputParameters = {
+  base: {
+    aliases: ['from', 'coin'],
+    type: 'string',
+    description: 'The symbol of symbols of the currency to query',
+    required: true,
+  },
+  quote: {
+    aliases: ['to', 'market'],
+    type: 'string',
+    description: 'The symbol of the currency to convert to',
+    required: true,
+  },
+}
+
+export type EndpointTypes = {
+  Request: {
+    Params: PriceEndpointParams
+  }
+  Response: SingleNumberResultResponse
+  CustomSettings: typeof customSettings
+  Provider: {
+    WsMessage: Message
+  }
+}
+
+export class BlocksizeWebsocketReverseMappingTransport<
+  T extends WebsocketTransportGenerics,
+  K,
+> extends WebsocketReverseMappingTransport<T, K> {
+  api_key = ''
+}
+
+export const websocketTransport: BlocksizeWebsocketReverseMappingTransport<EndpointTypes, string> =
+  new BlocksizeWebsocketReverseMappingTransport<EndpointTypes, string>({
+    url: ({ adapterConfig: { WS_API_ENDPOINT, API_KEY } }) => {
+      websocketTransport.api_key = API_KEY
+      return WS_API_ENDPOINT
+    },
+    handlers: {
+      open: (connection) => {
+        return new Promise((resolve, reject) => {
+          connection.on('message', (data: WebSocketRawData) => {
+            const parsed = JSON.parse(data.toString())
+            if (parsed.user_id) {
+              logger.info('Got logged in response, connection is ready')
+              resolve()
+            } else {
+              reject(new Error('Failed to make WS connection'))
+            }
+          })
+          const options = {
+            jsonrpc: '2.0',
+            method: 'authentication_logon',
+            params: { api_key: websocketTransport.api_key },
+          }
+          connection.send(JSON.stringify(options))
+        })
+      },
+      message: (message) => {
+        if (!('method' in message) || message.method !== 'vwap') return []
+        const [updates] = message.params.updates
+        const params = websocketTransport.getReverseMapping(updates.ticker)
+        if (!params) {
+          return []
+        }
+        if (!updates.price) {
+          logger.error(`The data provider didn't return any value`)
+          return []
```

Comment:
We can return an error here  so subsequent calls would just get the error from the cache. I was thinking something like what we do in `coingecko-test` [here](https://github.com/smartcontractkit/external-adapters-js/blob/59c4a2ee41116db7317915dda0dfe6063051ce70/packages/sources/coingecko-test/src/crypto-utils.ts#L88-L101)

Reply:
Done